src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilder.cs

```diff
@@ -66,11 +69,103 @@ public async IAsyncEnumerable<ProjectBuildResult> BuildProject()
+                            new ProjectBuildHandler(Logger, _analyzerConfiguration))
+                            {
+                                projectBuildHandler.ProjectAnalyzer = projectResult.ProjectAnalyzer;
+                                var result = projectBuildHandler.SyntaxOnlyBuild();
+                                ProjectResults.Add(result);
+                            }
+                        }
+                    }
+                } 
+                else
+                {
+                    builder.GenerateNoBuildAnalysis();
+
+                    var projectsInOrder = CreateDependencyQueue(projectReferencesMap);
+                    Dictionary<string, MetadataReference> references = new Dictionary<string, MetadataReference>();
+
+                    while (projectsInOrder.Count > 0)
+                    {
+                        var projectPath = projectsInOrder.Dequeue();
+                        var project = builder.Projects.Find(p => p.ProjectAnalyzer.ProjectFile.Path.Equals(projectPath));
+                        var projectReferencePaths = projectReferencesMap[projectPath]?.Distinct().ToHashSet<string>();
+
+                        using (ProjectBuildHandler projectBuildHandler =
+                            new ProjectBuildHandler(Logger, project.Project, _analyzerConfiguration))
+                        {
+                            projectBuildHandler.AnalyzerResult = project.AnalyzerResult;
+                            projectBuildHandler.ProjectAnalyzer = project.ProjectAnalyzer;
+                            var projectReferences = references.Where(r => projectReferencePaths.Contains(r.Key)).ToDictionary(p=>p.Key, p=> p.Value);
+                            var result = projectBuildHandler.SyntaxOnlyBuild(projectReferences);
+                            references.Add(projectPath, result.Compilation.ToMetadataReference());
+                            ProjectResults.Add(result);
+                        }
+                    }
+                }
+            }
+
+            return ProjectResults;
+        }
+
+        private Queue<string> CreateDependencyQueue(Dictionary<string, HashSet<string>> projectReferencesMap)
+        {
+            var projectsInOrder = new Queue<string>();
+            var builtProjects = new HashSet<string>();
+            foreach (var project in projectReferencesMap.Keys)
+            {
+                if (!builtProjects.Contains(project))
+                {
+                    CreateDependencyQueueHelper(project, builtProjects, projectReferencesMap, projectsInOrder);
+                }
+            }
+            return projectsInOrder;
+        }
+
+        private void CreateDependencyQueueHelper(string projectPath, HashSet<string> builtProjects, Dictionary<string, HashSet<string>> projectReferencesMap, Queue<string> buildOrder)
+        {
+            builtProjects.Add(projectPath);
+
+            foreach (var dependency in projectReferencesMap[projectPath])
```

Comment:
you can merge these two and move the dfs to one function, right ?

Reply:
If i do, i'll have to move some of the initialization code to the main function. Would rather keep this way for now. Let me know if there's something else in mind you're thinking

Reply:
no concern.. basically style of coding...